src/lib.rs

```diff
@@ -407,6 +380,39 @@ pub trait Rng: RngCore {
     fn sample<T, D: Distribution<T>>(&mut self, distr: D) -> T {
         distr.sample(self)
     }
+
+    /// Create an iterator that generates values using the given distribution.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use rand::{thread_rng, Rng};
+    /// use rand::distributions::{Alphanumeric, Range, Uniform};
+    ///
+    /// let mut rng = thread_rng();
+    ///
+    /// // Vec of 16 x f32:
+    /// let v: Vec<f32> = thread_rng().sample_iter(&Uniform).take(16).collect();
+    ///
+    /// // String:
+    /// let s: String = rng.sample_iter(&Alphanumeric).take(7).collect();
+    ///
+    /// // Combined values
+    /// println!("{:?}", thread_rng().sample_iter(&Uniform).take(5)
+    ///                              .collect::<Vec<(f64, bool)>>());
+    ///
+    /// // Dice-rolling:
+    /// let die_range = Range::new_inclusive(1, 6);
+    /// let mut roll_die = rng.sample_iter(&die_range);
+    /// while roll_die.next().unwrap() != 6 {
+    ///     println!("Not a 6; rolling again!");
+    /// }
+    /// ```
+    fn sample_iter<'a, T, D: Distribution<T>>(&'a mut self, distr: &'a D)
+        -> distributions::DistIter<'a, D, Self, T> where Self: Sized
```

Comment:
Interesting that this takes the distribution by reference while `sample` takes it by value. I'm not saying this is wrong. If we have both, then `distr.iter(&mut rng)` is still an option when passing by reference is required, so this could take the distribution by value (which would consume `die_range` above but only if the `&` is removed — we also implement `Distribution` for the references).

Reply:
I am not sure we can have both, because the iterator expects an reference instead of something owned. Or do we want to use some sort of `Cow` or `MaybeOwned` trick?

Reply:
I think it will work if you just change to `distr: D` here; essentially `Distribution::iter` receives a reference from this function stack.

Reply:
```
error[E0597]: `distr` does not live long enough
   --> src/lib.rs:414:9
    |
414 |         distr.sample_iter(self)
    |         ^^^^^ borrowed value does not live long enough
415 |     }
    |     - borrowed value only lives until here
    |
note: borrowed value must be valid for the lifetime 'a as defined on the method body at 411:5...
   --> src/lib.rs:411:5
    |
411 | /     fn sample_iter<'a, T, D: Distribution<T>>(&'a mut self, distr: D)
412 | |         -> distributions::DistIter<'a, D, Self, T> where Self: Sized
413 | |     {
414 | |         distr.sample_iter(self)
415 | |     }
    | |_____^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0597`.
```

Reply:
Sorry, of course the iterator is returned from `sample_iter`!